src/Compilers/CSharp/Portable/Binder/Binder_Query.cs

```diff
@@ -39,7 +39,7 @@ internal BoundExpression BindQuery(QueryExpressionSyntax node, DiagnosticBag dia
             }
 
             QueryTranslationState state = new QueryTranslationState();
-            state.fromExpression = MakeMemberAccessValue(boundFromExpression, diagnostics);
+            state.fromExpression = CheckValue(MakeMemberAccessValue(boundFromExpression, diagnostics), BindValueKind.RValueOrType, diagnostics);
```

Comment:
>CheckValue [](start = 35, length = 10)

Addition of this call here feels premature and inconsistent with how we deal with binding of a regular member access. Specifically, in case of ```TypeOrValueExpression```:
```
                    case BoundKind.TypeOrValueExpression:
                        {
                            // CheckValue call will occur in ReplaceTypeOrValueReceiver.
```  
It is quite possible that we are missing to call CheckValue at some point, but it feels like that should happen later in the pipeline. Addition of a new BindValueKind just to accommodate this call site feels suspicious and, for me, is one of the indicators of possibly doing work at wrong time/place.

Reply:
I'm not really sure where "later in the pipeline" would be here though. The next thing we do with `state.fromExpression` is potentially wrap it with a call to `Cast`, and it needs to have had default arguments/rvalueness checked by that point. In examining other uses of CheckValue, we call it once we know what context we're going to be using the value in, and this feels like the right to me. As to adding a new kind, most places in C# don't let you use a literal type as a receiver, but query syntax does, so it feels a bit special in that regard.

---
In reply to: [554076229](https://github.com/dotnet/roslyn/pull/50322#discussion_r554076229) [](ancestors = 554076229)

Reply:
> I'm not really sure where "later in the pipeline" would be here though. The next thing we do with state.fromExpression is potentially wrap it with a call to Cast, and it needs to have had default arguments/rvalueness checked by that point. In examining other uses of CheckValue, we call it once we know what context we're going to be using the value in, and this feels like the right to me. As to adding a new kind, most places in C# don't let you use a literal type as a receiver, but query syntax does, so it feels a bit special in that regard.

Query expressions in C# is a syntactic translation. The thing from ```from``` is a receiver, the ```Cast``` or any other query operator is a member access on that receiver. As pretty much anywhere in the language, a receiver of a member access is allowed to be a type, it doesn't feel like there is anything special about queries in that regard. My intuition is that process of binding that synthetic member access should pretty much do what ```BindMemberAccessWithBoundLeft``` does, or, perhaps, call it eventually. I am assuming that, if the query is manually translated into method calls, binding succeeds for that translated form, but fails for the untranslated form. For the problematic receiver, I would try to understand where binding for a query diverges from binding for a regular member access in terms of checking for default arguments/rvalueness for the receiver. That might give us some ideas.


---
In reply to: [554201536](https://github.com/dotnet/roslyn/pull/50322#discussion_r554201536) [](ancestors = 554201536,554076229)

Reply:
I've moved the checks into `MakeQueryInvocation`, where the rest of the related receiver checks occur. In particular, I discovered that we allow querying over type parameters, which should be invalid. An email to compat council has tentative support, but I'll hold off on merging until that's confirmed and a test insertion passes.

---
In reply to: [554217774](https://github.com/dotnet/roslyn/pull/50322#discussion_r554217774) [](ancestors = 554217774,554201536,554076229)